python/setup.py

```diff
@@ -0,0 +1,180 @@
+SCRIPTS_PATH = "%s/bin" % SPARK_HOME
+SCRIPTS_TARGET = "%s/bin" % TEMP_PATH
+JARS_TARGET = "%s/jars" % TEMP_PATH
+EXAMPLES_TARGET = "%s/examples" % TEMP_PATH
+
+# Check and see if we are under the spark path in which case we need to build the symlink farm.
+# This is important because we only want to build the symlink farm while under Spark otherwise we
+# want to use the symlink farm. And if the symlink farm exists under while under Spark (e.g. a
+# partially built sdist) we should error and have the user sort it out.
+in_spark = (os.path.isfile("../core/src/main/scala/org/apache/spark/SparkContext.scala") or
+            (os.path.isfile("../RELEASE") and len(glob.glob("../jars/spark*core*.jar")) == 1))
+
+if (in_spark):
+    # Construct links for setup
+    try:
+        os.mkdir(TEMP_PATH)
+    except:
+        print("Temp path for symlink to parent already exists %s" % TEMP_PATH, file=sys.stderr)
+        exit(-1)
+
+try:
+    if (in_spark):
+        # Construct the symlink farm - this is necessary since we can't refer to the path above the
+        # package root and we need to copy the jars and scripts which are up above the python root.
+        if getattr(os, "symlink", None) is not None:
+            os.symlink(JARS_PATH, JARS_TARGET)
+            os.symlink(SCRIPTS_PATH, SCRIPTS_TARGET)
+            os.symlink(EXAMPLES_PATH, EXAMPLES_TARGET)
+        else:
+            # For windows fall back to the slower copytree
+            copytree(JARS_PATH, JARS_TARGET)
+            copytree(SCRIPTS_PATH, SCRIPTS_TARGET)
+            copytree(EXAMPLES_PATH, EXAMPLES_TARGET)
+    else:
+        # If we are not inside of SPARK_HOME verify we have the required symlink farm
+        if not os.path.exists(JARS_TARGET):
+            print("To build packaging must be in the python directory under the SPARK_HOME.",
+                  file=sys.stderr)
+        # We copy the shell script to be under pyspark/python/pyspark so that the launcher scripts
+        # find it where expected. The rest of the files aren't copied because they are accessed
+        # using Python imports instead which will be resolved correctly.
+        try:
+            os.makedirs("pyspark/python/pyspark")
+        except OSError:
+            # Don't worry if the directory already exists.
+            True
+        copyfile("pyspark/shell.py", "pyspark/python/pyspark/shell.py")
+
+    if not os.path.isdir(SCRIPTS_TARGET):
+        print("You must first create a source dist and install that source dist.", file=sys.stderr)
+        exit(-1)
+
+    # Scripts directive requires a list of each script path and does not take wild cards.
+    script_names = os.listdir(SCRIPTS_TARGET)
+    scripts = map(lambda script: os.path.join(SCRIPTS_TARGET, script), script_names)
+    # We add find_spark_home.py to the bin directory we install so that pip installed PySpark
+    # will search for SPARK_HOME with Python.
+    scripts.append("pyspark/find_spark_home.py")
```

Comment:
This line causes error in python3.

    Traceback (most recent call last):
      File "setup.py", line 105, in <module>
        scripts.append("pyspark/find_spark_home.py")
        AttributeError: 'map' object has no attribute 'append'